### ISSUE TITLE
Start static pod take long time if APIServer down

### ISSUE THREAD

**Issue by TeddyAndrieux** [kubernetes/kubernetes#103658]
#### What happened:

I have a kubelet configured to connect to a local APIServer that run as a static pod managed by kubelet, if this APIServer goes down or get edited and I have a bunch of Pod (not static ones) that use PVCs kubelet will take a lot of time to schedule APIServer static pod again.

So in kubelet logs I have a bunch of
```
Jul 07 10:45:21 HP-DL380-01_21014 kubelet[15723]: E0707 10:45:21.318938   15723 kubelet.go:1684] Unable to attach or mount volumes for pod "kube-apiserver-bootstrap_kube-system(c4786b9c39e0789c7ef802f731bf115b)": unmounted volumes=[k8s-encryption etc-pki k8s-certs ca-certs], unattached volumes=[k8s-encryption etc-pki k8s-certs ca-certs]: timed out waiting for the condition; skipping pod
Jul 07 10:45:21 HP-DL380-01_21014 kubelet[15723]: E0707 10:45:21.318963   15723 pod_workers.go:191] Error syncing pod c4786b9c39e0789c7ef802f731bf115b ("kube-apiserver-bootstrap_kube-system(c4786b9c39e0789c7ef802f731bf115b)"), skipping: unmounted volumes=[k8s-encryption etc-pki k8s-certs ca-certs], unattached volumes=[k8s-encryption etc-pki k8s-certs ca-certs]: timed out waiting for the condition
```

#### What you expected to happen:

Static pods to be scheduled properly and in a short amount of time even if APIServer is down

#### How to reproduce it (as minimally and precisely as possible):

Basically a simple Kubernetes deployment with a bunch of Pods on a specific node with a bunch PVCs then just edit static manifest of APIServer

#### Anything else we need to know?:

I investigated it a bit and it seems it's due to kubelet that iterate on all pods the same way, no matter if it's a static pod or not and also if the Pod kubelet want to deploy has a PVC he will try to get PVC information from APIServer (that may be down and so it will timeout).

To me, kubelet should either, first try to schedule all static Pods before doing other Pods, or handle static Pods and "classic" Pods in 2 different loops in parallel.

"""Workaround""": As a simple "workaround" to just make APIServer come back a bit faster, I did restart kubelet

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.6 (but also tested with 1.21.2 and we have the same issue)
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): CentOS 7.9.2009
- Kernel (e.g. `uname -a`): 3.10.0-1160.el7.x86_64
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by ehashman** [comment]
This is an issue with your architecture. If your static pod depends on a PVC, it will necessarily need to talk to the API server in order to provision and bind the volume. You can read more in the docs: https://kubernetes.io/docs/concepts/storage/persistent-volumes/

Static pods are not recommended for general use and you should not try to treat them like regular pods.

In any case, Kubernetes does not use issues on this repo for support requests. If you have a question on how to use Kubernetes or to debug a specific issue, please visit our [forums](https://discuss.kubernetes.io/).

/sig storage
/remove-kind bug
/kind support
/close

**Comment by TeddyAndrieux** [comment]
It does NOT depend on a PVC.

As it's already mentioned in my initial message, I have static pods only for "classic mandatory Kubernetes pods" so basically APIServer etcd, scheduler, and controller manager that does NOT have any PVC. 
But those pods take a long time to be scheduled because I have a bunch of Pod (non-static pod) that have some PVC and so kubelet iter on all those pods and try to retrieve PVC information for those "NON STATIC PODS" so it needs to first timeout for all those Pods before actually schedule my static pod that does not have any PVC

It's not a support request

/reopen
/remove-kind support
/kind bug

**Comment by ehashman** [comment]
@TeddyAndrieux please attach full kubelet logs to the bug demonstrating this, thanks!

/triage needs-information

**Comment by TeddyAndrieux** [comment]
I already put some lines, I do not have much since kubelet does not logs when fetching PVC but basically I added some logs to debug and it hang on every static Pods when fetching the PVC
```go
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:633:// getPVCExtractPV fetches the PVC object with the given namespace and name from
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-634-// the API server, checks whether PVC is being deleted, extracts the name of the PV
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-635-// it is pointing to and returns it.
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-636-// An error is returned if the PVC object's phase is not "Bound".
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:637:func (dswp *desiredStateOfWorldPopulator) getPVCExtractPV(
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-638-    namespace string, claimName string) (*v1.PersistentVolumeClaim, error) {
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-639-    pvc, err :=
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-640-            dswp.kubeClient.CoreV1().PersistentVolumeClaims(namespace).Get(context.TODO(), claimName, metav1.GetOptions{})
```

And on my tests, it took like 3 minutes to timeout to get this PVC (and during this time static pod volumes, which are not PVC, does not get bind, so apiserver cannot be scheduled, that's why I have this log in kubelet)

```
Jul 07 10:45:21 HP-DL380-01_21014 kubelet[15723]: E0707 10:45:21.318938   15723 kubelet.go:1684] Unable to attach or mount volumes for pod "kube-apiserver-bootstrap_kube-system(c4786b9c39e0789c7ef802f731bf115b)": unmounted volumes=[k8s-encryption etc-pki k8s-certs ca-certs], unattached volumes=[k8s-encryption etc-pki k8s-certs ca-certs]: timed out waiting for the condition; skipping pod
Jul 07 10:45:21 HP-DL380-01_21014 kubelet[15723]: E0707 10:45:21.318963   15723 pod_workers.go:191] Error syncing pod c4786b9c39e0789c7ef802f731bf115b ("kube-apiserver-bootstrap_kube-system(c4786b9c39e0789c7ef802f731bf115b)"), skipping: unmounted volumes=[k8s-encryption etc-pki k8s-certs ca-certs], unattached volumes=[k8s-encryption etc-pki k8s-certs ca-certs]: timed out waiting for the condition
```


So to explain a bit more the problem I faced, the function `processPodVolumes`
```go
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:308:// processPodVolumes processes the volumes in the given pod and adds them to the
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-309-// desired state of the world.
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:310:func (dswp *desiredStateOfWorldPopulator) processPodVolumes(
```
Loop on all Pods with volumes (and not only pods with PVC, it also include static Pod with some "hostPath volumes").
For each pod volume, it call `createVolumeSpec`
```go
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:336:                    dswp.createVolumeSpec(podVolume, pod, mounts, devices)
```
```go
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:507:// createVolumeSpec creates and returns a mutable volume.Spec object for the
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-508-// specified volume. It dereference any PVC to get PV objects, if needed.
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-509-// Returns an error if unable to obtain the volume at this time.
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:510:func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(
```

And if this volume then it tries to get the PVC from APIServer
```go
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:534:            pvc, err := dswp.getPVCExtractPV(
pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go-535-                    pod.Namespace, pvcSource.ClaimName)
```
So it hangs because APIServer is not ready because APIServer static Pod has some volume that need to be bind, so `processPodVolumes` need to iterate on APIServer for this.

IMHO, kubelet should have 2 differents loops one for static pod and one for others pods basically 2 `processPodVolumes` but one that focus on the static pods (that, as you mention, should not have any PVC)

**Comment by ehashman** [comment]
@TeddyAndrieux I need the full logs (v=2 fine, v=4 even better) from the your kubelet in order to confirm this and reproduce against kubernetes HEAD. You have stated that the issue is with the kubelet PLEG and it hanging on PVCs before it does another sync loop for the static pod, but you have not attached a kubelet log that demonstrates that.

**Comment by TeddyAndrieux** [comment]
Ok so I spend some time to reproduce (and describe how to reproduce) this issue (and also put some logs)

In my case, it's because I use a "proxy" for APIServer so that I have some load balancing on APIServer and also fallback on non-local APIServer if needed (basically non-Control Plane node kubelet hit different APIServers and Control Plane node can fallback on distant APIServer if local one is not available). But to me it can happen in other situations and anyway it does not make real sense to me to have a single loop for static Pods and non-static Pods (as I already described in previous comments)

#### Reproduce

- Deploy a classic kubeadm cluster + CNI (in my case I used kubeadm 1.21.3 and Calico)
- <details><summary>Deploy a simple APIServer proxy</summary>

  ```yaml
  # cat /etc/kubernetes/manifests/apiserver-proxy.yaml 
  apiVersion: v1
  kind: Pod
  metadata:
    name: apiserver-proxy
    namespace: kube-system
    labels:
      addonmanager.kubernetes.io/mode: Reconcile
      k8s-app: apiserver-proxy
      app: apiserver-proxy
      app.kubernetes.io/name: apiserver-proxy
  spec:
    hostNetwork: true
    dnsPolicy: ClusterFirstWithHostNet
    nodeSelector:
      kubernetes.io/os: linux
    securityContext:
      readOnlyRootFilesystem: true
      allowPrivilegeEscalation: false
      capabilities:
        drop:
          - ALL
    priorityClassName: system-node-critical
    containers:
    - name: apiserver-proxy
      image: nginx:stable-alpine
      imagePullPolicy: IfNotPresent
      ports:
      - name: https
        containerPort: 7443
      - name: healthcheck
        containerPort: 7080
      resources:
        requests:
          cpu: 25m
          memory: 32M
      livenessProbe:
        httpGet:
          path: /healthz
          port: healthcheck
          host: 127.0.0.1
      readinessProbe:
        httpGet:
          path: /healthz
          port: healthcheck
          host: 127.0.0.1
      volumeMounts:
      - mountPath: /etc/nginx/nginx.conf
        name: nginx-config
        readOnly: true
    volumes:
    - name: nginx-config
      hostPath:
        path: /var/lib/apiserver-proxy.conf
        type: File
  ```
  ```nginx
  # cat /var/lib/apiserver-proxy.conf 
  error_log stderr notice;

  worker_processes 2;
  worker_rlimit_nofile 130048;
  worker_shutdown_timeout 10s;

  events {
    multi_accept on;
    use epoll;
    worker_connections 16384;
  }


  stream {
    upstream apiserver {
      least_conn;
      # Local server
      server 10.200.7.30:6443 weight=100;
      # other APIServers
      # server 10.10.10.10:6443 weight=1;
    }

    server {
      listen        7443;
      proxy_pass    apiserver;
      proxy_timeout 10m;
      proxy_connect_timeout 1s;
    }
  }

  http {
    aio threads;
    aio_write on;
    tcp_nopush on;
    tcp_nodelay on;

    keepalive_timeout 10m;
    keepalive_requests 100;
    reset_timedout_connection on;
    autoindex off;

    server {
      listen 127.0.0.1:7080;
      location /healthz {
        access_log off;
        return 200 'ok';
      }
    }
  }
  ```
  </details>

- <details><summary>Edit kubelet kubeconfig to use this APIServer proxy</summary>

  First make sure this APIServer proxy works well
  ```console
  # kubectl --kubeconfig=/etc/kubernetes/kubelet.conf --server="https://10.200.7.30:7443" get node
  NAME                         STATUS   ROLES                  AGE     VERSION
  reproduce-103658.novalocal   Ready    control-plane,master   4h34m   v1.21.3
  ```
  Edit "/etc/kubernetes/kubelet.conf` so that we use APIServer proxy `7443`
  ```console
  # grep "server" /etc/kubernetes/kubelet.conf 
      server: https://10.200.7.30:7443
  ```
  </details>

- Create a simple Storage Class
- <details><summary>Create a bunch of PV</summary>

  In my case I used LVM
  ```yaml
  apiVersion: v1
  kind: PersistentVolume
  metadata:
    name: pv-@N@
  spec:
    capacity:
      storage: 5Gi
    volumeMode: Filesystem
    accessModes:
    - ReadWriteOnce
    storageClassName: sc-0
    local:
      path: /dev/my-vg/vol-@N@
    nodeAffinity:
      required:
        nodeSelectorTerms:
        - matchExpressions:
          - key: kubernetes.io/hostname
            operator: In
            values:
            - reproduce-103658.novalocal
  ```
  ```console
  # for i in `seq 1 50`; do lvcreate -L 5Gi -n vol-$i my-vg; sed "s/@N@/$i/g" pv.yaml | kubectl apply -f -; done
    Logical volume "vol-1" created.
  persistentvolume/pv-1 created
  [...]
    Logical volume "vol-50" created.
  persistentvolume/pv-50 created
  ```
  </details>

- <details><summary>Create a bunch of PVC and Pods using those PV</summary>

  In my case I deployed Pods that just sleep but it's the same with "real Pods" that use PVC
  ```yaml
  apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    name: pvc-@N@
  spec:
    accessModes:
      - ReadWriteOnce
    resources:
      requests:
        storage: 5Gi
    storageClassName: sc-0

  ---
  apiVersion: v1
  kind: Pod
  metadata:
    name: pod-@N@
  spec:
    containers:
    - image: busybox
      command:
        - "sleep"
        - "36000"
      name: sleep-container
      volumeMounts:
      - mountPath: /var/tmp/abc
        name: my-vol
    restartPolicy: Always
    volumes:
    - name: my-vol
      persistentVolumeClaim:
        claimName: pvc-@N@
  ``` 
  ```console
  # for i in `seq 1 50`; do sed "s/@N@/$i/g" pod.yaml | kubectl apply -f -; done
  persistentvolumeclaim/pvc-1 created
  pod/pod-1 created
  [...]
  persistentvolumeclaim/pvc-50 created
  pod/pod-50 created
  ```
  </details>

- Edit APIServer manifest to ... add a new label or annotation or ....whatever

From time to time APIServer will take ages to restart because it will retry to get PVC

On the tests I did it took like ~20 minutes to restart and depending on .... luck it may take less time or way longer (on a previous test I did it was like 1hour)

From [log.txt](https://github.com/kubernetes/kubernetes/files/6848989/log.txt)

```
Jul 20 13:48:15 reproduce-103658.novalocal kubelet[16811]: I0720 13:48:15.462511   16811 config.go:383] "Receiving a new pod" pod="kube-system/kube-apiserver-reproduce-103658.novalocal"
```

And we see volume does not get bind
```
Jul 20 13:50:46 reproduce-103658.novalocal kubelet[16811]: E0720 13:50:46.331319   16811 kubelet.go:1701] "Unable to attach or mount volumes for pod; skipping pod" err="unmounted volumes=[k8s-certs ca-certs etc-pki], unattached volumes=[k8s-certs ca-certs etc-pki]: timed out waiting for the condition" pod="kube-system/kube-apiserver-reproduce-103658.novalocal"
Jul 20 13:53:04 reproduce-103658.novalocal kubelet[16811]: E0720 13:53:04.333796   16811 kubelet.go:1701] "Unable to attach or mount volumes for pod; skipping pod" err="unmounted volumes=[k8s-certs ca-certs etc-pki], unattached volumes=[k8s-certs ca-certs etc-pki]: timed out waiting for the condition" pod="kube-system/kube-apiserver-reproduce-103658.novalocal"
Jul 20 13:57:37 reproduce-103658.novalocal kubelet[16811]: E0720 13:57:37.971794   16811 kubelet.go:1701] "Unable to attach or mount volumes for pod; skipping pod" err="unmounted volumes=[ca-certs etc-pki k8s-certs], unattached volumes=[ca-certs etc-pki k8s-certs]: timed out waiting for the condition" pod="kube-system/kube-apiserver-reproduce-103658.novalocal"
```

And this one gets scheduled only 20 minutes later because of this fetch of PVC for non-static Pod

```
Jul 20 14:07:31 reproduce-103658.novalocal kubelet[16811]: I0720 14:07:31.108213   16811 desired_state_of_world_populator.go:349] "Added volume to desired state" pod="kube-system/kube-apiserver-reproduce-103658.novalocal" volumeName="ca-certs" volumeSpecName="ca-certs"
Jul 20 14:07:31 reproduce-103658.novalocal kubelet[16811]: I0720 14:07:31.108238   16811 desired_state_of_world_populator.go:349] "Added volume to desired state" pod="kube-system/kube-apiserver-reproduce-103658.novalocal" volumeName="etc-pki" volumeSpecName="etc-pki"
Jul 20 14:07:31 reproduce-103658.novalocal kubelet[16811]: I0720 14:07:31.108258   16811 desired_state_of_world_populator.go:349] "Added volume to desired state" pod="kube-system/kube-apiserver-reproduce-103658.novalocal" volumeName="k8s-certs" volumeSpecName="k8s-certs"
Jul 20 14:07:31 reproduce-103658.novalocal kubelet[16811]: I0720 14:07:31.468728   16811 prober.go:125] "Probe succeeded" probeType="Readiness" pod="kube-system/kube-apiserver-reproduce-103658.novalocal" podUID=aab555052b36551675038b782d59201e containerName="kube-apiserver"
```

When APIServer is down
```console
# kubectl --kubeconfig=/etc/kubernetes/kubelet.conf get node -v=5
I0720 13:48:47.193664   17224 cert_rotation.go:137] Starting client certificate rotation controller
I0720 13:48:47.202387   17224 request.go:959] Got a Retry-After 1s response for attempt 1 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:48.203313   17224 request.go:959] Got a Retry-After 1s response for attempt 2 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:49.204651   17224 request.go:959] Got a Retry-After 1s response for attempt 3 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:50.205982   17224 request.go:959] Got a Retry-After 1s response for attempt 4 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:51.207004   17224 request.go:959] Got a Retry-After 1s response for attempt 5 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:52.208027   17224 request.go:959] Got a Retry-After 1s response for attempt 6 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:53.209592   17224 request.go:959] Got a Retry-After 1s response for attempt 7 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:54.210660   17224 request.go:959] Got a Retry-After 1s response for attempt 8 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:55.211610   17224 request.go:959] Got a Retry-After 1s response for attempt 9 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:56.212514   17224 request.go:959] Got a Retry-After 1s response for attempt 10 to https://10.200.7.30:7443/api/v1/nodes?limit=500
I0720 13:48:57.214452   17224 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "an error on the server (\"\") has prevented the request from succeeding (get nodes)",
  "reason": "InternalError",
  "details": {
    "kind": "nodes",
    "causes": [
      {
        "reason": "UnexpectedServerResponse"
      }
    ],
    "retryAfterSeconds": 1
  },
  "code": 500
}]
Error from server (InternalError): an error on the server ("") has prevented the request from succeeding (get nodes)
```

#### Reason

Basically, kubelet does some retry when getting PVC because my APIServer proxy does not refuse the connection, but we could have the same issue for any reason that makes retrieving the PVC timeout or retries etc ...

To me, it's really an issue that we have static Pod volume treated in the same "loop" as other "non-static Pods" because it means "non-static Pods" impact scheduling of static Pods

**Comment by TeddyAndrieux** [comment]
IMHO this issue is still valid

/remove-lifecycle stale

**Comment by gjkim42** [comment]
> Basically, kubelet does some retry when getting PVC because my APIServer proxy does not refuse the connection, but we could have the same issue for any reason that makes retrieving the PVC timeout or retries etc ... And IMO those retrieves to get PVC make sense just it shouldn't affect static Pod scheduling.

Not sure how to solve getting PVC failing issue, but we could remove the dependency of the non-static pod loop on the static pod loop.

> To me, it's really an issue that we have static Pod volume treated in the same "loop" as other "non-static Pods" because it means "non-static Pods" impact scheduling of static Pods

Agree. I think we should split the main loop into two loops(static pod loop and non-static pod loop).

@smarterclayton @ehashman 

What do you think?

/assign

**Comment by gjkim42** [comment]
/unassign
(lack of resources...)

**Comment by TeddyAndrieux** [comment]
To me this issue is still there in recent versions 

/reopen
